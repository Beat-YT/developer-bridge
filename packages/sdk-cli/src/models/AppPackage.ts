import jszip from 'jszip';
import lodash from 'lodash';
import { RawSourceMap } from 'source-map';

import mapValues from '../util/mapValues';

interface PackageComponents {
  device: {
    [family: string]: {
      artifact: Buffer,
      platform?: string[],
    },
  };
  companion?: Buffer;
}

interface DeviceSourceMaps {
  [family: string]: ComponentSourceMaps;
}

export interface SourceMaps {
  device: DeviceSourceMaps;
  companion?: ComponentSourceMaps;
  settings?: ComponentSourceMaps;
}

export interface ComponentSourceMaps {
  [filename: string]: RawSourceMap;
}

interface BundledComponentSourceMaps {
  [filename: string]: string;
}

interface ManifestSourceMaps {
  device: {
    [family: string]: BundledComponentSourceMaps;
  };
  companion?: BundledComponentSourceMaps;
  settings?: BundledComponentSourceMaps;
}

interface ManifestCommon {
  manifestVersion: number;
  buildId: string;
  appId: string;
  requestedPermissions: string[];
  components: {
    companion?: {};
  };
}

interface ManifestV5 extends ManifestCommon {
  manifestVersion: 5;
  components: {
    watch: string;
    companion?: string;
  };
  platform: string[];
}

interface ManifestV6 extends ManifestCommon {
  manifestVersion: 6;
  sdkVersion: SDKVersion;
  components: {
    watch: {
      [family: string]: {
        platform?: string[];
        filename: string;
      };
    };
    companion?: {
      filename: string;
    };
  };
  sourceMaps?: ManifestSourceMaps;
}

interface SDKVersion {
  deviceApi: string;
  companionApi?: string;
}

function getFile(zip: jszip, path: string) {
  const file = zip.file(path);
  if (!file || file.dir) throw new Error(`${path} not present in zip file`);
  return file;
}

async function getBufferFromZip(zip: jszip, path: string) {
  return getFile(zip, path).async('nodebuffer');
}

async function getTextFromZip(zip: jszip, path: string) {
  return getFile(zip, path).async('text');
}

function packageComponentSourceMaps(files: ComponentSourceMaps, prefix: string, zip: jszip) {
  return lodash.mapValues(files, (sourceMap, file) => {
    const zipFileName = `sourceMaps/${prefix}/${file}.map`;
    zip.file(zipFileName, JSON.stringify(sourceMap));
    return zipFileName;
  });
}

function bundleSourceMaps(zip: jszip, sourceMaps?: SourceMaps) {
  if (!sourceMaps) return;

  const sourceMapObject: ManifestSourceMaps = { device: {} };
  const packageSourceMaps = (files: ComponentSourceMaps, prefix: string) =>
    packageComponentSourceMaps(files, prefix, zip);

  Object.entries(sourceMaps.device).forEach(([family, files]) => {
    sourceMapObject.device[family] = packageSourceMaps(files, `device/${family}`);
  });

  if (sourceMaps.companion) {
    sourceMapObject.companion = packageSourceMaps(sourceMaps.companion, 'companion');
  }

  if (sourceMaps.settings) {
    sourceMapObject.settings = packageSourceMaps(sourceMaps.settings, 'settings');
  }

  return sourceMapObject;
}

const extractComponentSourceMaps =
  (sourceMapPaths: BundledComponentSourceMaps, zip: jszip): Promise<ComponentSourceMaps> =>
    mapValues(sourceMapPaths, path => getTextFromZip(zip, path).then(JSON.parse));

async function extractSourceMaps(zip: jszip, sourceMapManifest?: ManifestSourceMaps) {
  if (!sourceMapManifest) return undefined;

  const extractComponent = (component?: BundledComponentSourceMaps) =>
    component && extractComponentSourceMaps(component, zip);

  return {
    device: await mapValues(
      sourceMapManifest.device,
      component => extractComponentSourceMaps(component, zip),
    ) as DeviceSourceMaps,
    companion: await extractComponent(sourceMapManifest.companion),
    settings: await extractComponent(sourceMapManifest.settings),
  };
}

abstract class ManifestParserBase {
  protected manifest!: ManifestCommon;

  abstract getSourceMapExtractor(): (zip: jszip) => Promise<SourceMaps | undefined>;

  pullMetadata() {
    return {
      buildId: this.manifest.buildId,
      uuid: this.manifest.appId,
      requestedPermissions: this.manifest.requestedPermissions,
    };
  }

  getSDKVersions() {
    const versions: SDKVersion = {
      deviceApi: '1.0.0',
    };
    if (this.manifest.components.companion) versions.companionApi = '1.0.0';
    return versions;
  }
}

class ManifestParserV5 extends ManifestParserBase {
  constructor(protected manifest: ManifestV5) {
    super();
  }

  getDeviceComponents() {
    if (typeof this.manifest.components !== 'object' || !this.manifest.components.watch) {
      throw new Error('No components listed in manifest.json');
    }

    if (!Array.isArray(this.manifest.platform)) {
      throw new Error('Missing platform descriptors');
    }

    return this.manifest.platform.map<[string, { filename: string, platform?: string[] }]>(
      (platformDescriptor) => {
        const [, family, platform] = /^([^:]+):?(.+)?$/.exec(platformDescriptor)!;

        return [
          family.toLowerCase(),
          {
            platform: platform ? [platform] : undefined,
            filename: this.manifest.components.watch,
          },
        ];
      },
    );
  }

  getCompanionFilename() {
    return this.manifest.components.companion;
  }

  getSourceMapExtractor() {
    return () => Promise.resolve(undefined);
  }
}

class ManifestParserV6 extends ManifestParserBase {
  constructor(protected manifest: ManifestV6) {
    super();
  }

  getDeviceComponents() {
    if (typeof this.manifest.components !== 'object') {
      throw new Error('No components listed in manifest.json');
    }

    if (!this.manifest.components.watch) return [];

    return Object.entries(this.manifest.components.watch);
  }

  getCompanionFilename() {
    return lodash.get(this.manifest, 'components.companion.filename');
  }

  getSDKVersions() {
    return {
      ...super.getSDKVersions(),
      ...this.manifest.sdkVersion,
    };
  }

  getSourceMapExtractor() {
    return (zip: jszip) => extractSourceMaps(zip, this.manifest.sourceMaps);
  }
}

function getManifestParser(manifest: ManifestCommon) {
  switch (manifest.manifestVersion) {
    case 5:
      return new ManifestParserV5(manifest as ManifestV5);
    case 6:
      return new ManifestParserV6(manifest as ManifestV6);
    default:
      throw new Error(`Unsupported manifest version ${manifest.manifestVersion}`);
  }
}

export default class AppPackage {
  public buildId: string;
  public components: PackageComponents;
  public sourceMaps?: SourceMaps;
  public requestedPermissions: string[];
  public uuid: string;
  public sdkVersion: SDKVersion;

  /** Used by Studio's sideload view to show the manifest content
   * @deprecated  */
  public manifest?: ManifestV5 | ManifestV6;

  constructor({
    buildId,
    components,
    sourceMaps,
    requestedPermissions,
    uuid,
    sdkVersion,
  }: {
    buildId: string,
    components: PackageComponents,
    sourceMaps?: SourceMaps,
    requestedPermissions: string[],
    uuid: string,
    sdkVersion: SDKVersion,
  }) {
    this.buildId = buildId;
    this.components = components;
    this.sourceMaps = sourceMaps;
    this.requestedPermissions = requestedPermissions;
    this.uuid = uuid;
    this.sdkVersion = sdkVersion;
  }

  generateArtifact() {
    const zip = new jszip();

    const manifest: ManifestV6 = {
      manifestVersion: 6,
      buildId: this.buildId,
      appId: this.uuid,
      requestedPermissions: this.requestedPermissions,
      components: { watch: {} },
      sdkVersion: this.sdkVersion,
      sourceMaps: bundleSourceMaps(zip, this.sourceMaps),
    };

    for (const [family, { platform, artifact }] of Object.entries(this.components.device)) {
      const filename = `device-${family}.zip`;
      manifest.components.watch[family] = { platform, filename };
      zip.file(filename, artifact);
    }

    if (this.components.companion) {
      manifest.components.companion = {
        filename: 'companion.zip',
      };
      zip.file(manifest.components.companion.filename, this.components.companion);
    }

    zip.file('manifest.json', JSON.stringify(manifest, null, 2));
    return zip.generateAsync({ type: 'nodebuffer' });
  }

  static async fromArtifact(artifactData: Buffer) {
    const fbaZip = await jszip.loadAsync(artifactData);
    const textFile = (path: string) => getTextFromZip(fbaZip, path);
    const bufferFile = (path: string) => getBufferFromZip(fbaZip, path);

    const manifestJSON = JSON.parse(await textFile('manifest.json'));
    const parser = getManifestParser(manifestJSON);

    const device = await Promise.all(
      parser.getDeviceComponents().map(
        ([family, { platform, filename }]) => bufferFile(filename)
          .then(artifact => [family, { platform, artifact }]),
      ),
    ).then(lodash.fromPairs);

    const companionFilename = parser.getCompanionFilename();
    const companion = companionFilename ? await bufferFile(companionFilename) : undefined;

    const sourceMaps = await parser.getSourceMapExtractor()(fbaZip);

    const appPackage = new this({
      ...parser.pullMetadata(),
      sourceMaps,
      components: {
        device,
        companion,
      },
      sdkVersion: parser.getSDKVersions(),
    });

    appPackage.manifest = manifestJSON;

    return appPackage;
  }
}
