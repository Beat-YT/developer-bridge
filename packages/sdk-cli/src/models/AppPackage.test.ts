import jszip from 'jszip';

import AppPackage, { SourceMaps } from './AppPackage';

const manifestCommon = {
  appId: 'eff9d309-eadd-41d7-9af3-696eafc3fa31',
  buildId: '0x0123456789abcdef',
};

const testPackage = {
  buildId: manifestCommon.buildId,
  uuid: manifestCommon.appId,
  components: {
    device: {
      higgs: {
        platform: ['1.2.3+'],
        artifact: Buffer.alloc(0),
      },
    },
    companion: Buffer.alloc(0),
  },
  requestedPermissions: ['access_activity', 'access_location'],
  sdkVersion: {
    deviceApi: '3.0.0',
    companionApi: '2.5.0',
  },
};

const rawSourceMap = {
  version: 3,
  sources: ['app/index.js'],
  names: [],
  mappings: 'someMappings',
  file: 'index.js',
};

describe('fromArtifact', () => {
  const manifestV5 = {
    ...manifestCommon,
    manifestVersion: 5,
    platform: [],
  };

  const manifestV6 = {
    ...manifestCommon,
    manifestVersion: 6,
  };

  type FilesObject = { [path: string]: object | string };

  function buildPackage(files: FilesObject = {}) {
    const zip = new jszip();
    for (const [path, content] of Object.entries(files)) {
      zip.file(path, typeof content === 'object' ? JSON.stringify(content) : content);
    }
    return zip.generateAsync({ type: 'nodebuffer' });
  }

  function itRejects(name: string, files: FilesObject = {}) {
    it(`rejects ${name}`, async () =>
      expect(AppPackage.fromArtifact(await buildPackage(files)))
        .rejects.toMatchSnapshot());
  }

  function itAccepts(name: string, files: FilesObject) {
    it(`accepts ${name}`, async () =>
      expect(AppPackage.fromArtifact(await buildPackage(files)))
        .resolves.toMatchSnapshot());
  }

  it('rejects a non-zip file', () =>
    expect(AppPackage.fromArtifact('' as any)).rejects.toMatchSnapshot());

  itRejects('a zip file without a manifest');

  itRejects('a manifest that is not JSON', {
    'manifest.json': 'Not JSON',
  });

  itRejects('a manifest with unsupported version', {
    'manifest.json': { manifestVersion: 100 },
  });

  [5, 6].forEach((version) => {
    itRejects(`a v${version} manifest with mistyped components`, {
      'manifest.json': {
        manifestVersion: version,
        components: 'components!',
      },
    });
  });

  itRejects('a v5 manifest with no device component', {
    'manifest.json': {
      ...manifestV5,
      components: { companion: 'companion.zip' },
    },
    'companion.zip': '',
  });

  itRejects('a v5 manifest which references a nonexistent device component file', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS'],
      components: { watch: 'doesnotexist.zip' },
    },
  });

  itRejects('a v6 manifest which references a nonexistent device component file', {
    'manifest.json': {
      ...manifestV6,
      components: { watch: {
        higgs: { filename: 'doesnotexist.zip' },
      } },
    },
  });

  itRejects('a v5 manifest which references a nonexistent companion component file', {
    'manifest.json': {
      ...manifestV5,
      components: {
        watch: 'device.zip',
        companion: 'companion.zip',
      },
    },
    'device.zip': '',
  });

  itRejects('a v6 manifest which references a nonexistent companion component file', {
    'manifest.json': {
      ...manifestV6,
      components: {
        watch: {
          higgs: { filename: 'higgs.zip' },
        },
        companion: { filename: 'companion.zip' },
      },
    },
    'higgs.zip': '',
  });

  itRejects('a v5 manifest with no platform descriptor', {
    'manifest.json': {
      ...manifestV5,
      platform: undefined,
      components: { watch: 'device.zip' },
    },
    'device.zip': '',
  });

  itAccepts('a v5 manifest with an unversioned platform', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS'],
      components: { watch: 'device.zip' },
    },
    'device.zip': 'device.zip contents',
  });

  itAccepts('a v5 manifest with a versioned platform', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS:32.1.16+'],
      components: { watch: 'device.zip' },
    },
    'device.zip': 'device.zip contents',
  });

  itAccepts('a v5 manifest with multiple platforms', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS:32.1.16+', 'MESON:1.2.3+'],
      components: { watch: 'device.zip' },
    },
    'device.zip': 'device.zip contents',
  });

  itAccepts('a v6 manifest with no platform spec for the device', {
    'manifest.json': {
      ...manifestV6,
      components: {
        watch: {
          higgs: { filename: 'higgs.zip' },
        },
      },
    },
    'higgs.zip': 'higgs.zip contents',
  });

  itAccepts('a v6 manifest with multiple device components', {
    'manifest.json': {
      ...manifestV6,
      components: {
        watch: {
          higgs: {
            platform: ['23.1.5', '32.4.16+'],
            filename: 'higgs.zip',
          },
          meson: {
            platform: ['32.4.17+'],
            filename: 'meson.zip',
          },
        },
      },
    },
    'higgs.zip': 'higgs',
    'meson.zip': 'meson',
  });

  itAccepts('a v5 manifest with a companion', {
    'manifest.json': {
      ...manifestV5,
      platform: ['HIGGS'],
      components: {
        watch: 'device.zip',
        companion: 'companion.zip',
      },
    },
    'device.zip': 'device',
    'companion.zip': 'companion',
  });

  itAccepts('a v6 manifest with a companion', {
    'manifest.json': {
      ...manifestV6,
      components: {
        watch: { higgs: { filename: 'higgs.zip' } },
        companion: { filename: 'companion.zip' },
      },
    },
    'higgs.zip': 'higgs',
    'companion.zip': 'companion',
  });

  itAccepts('a v6 manifest with no device component', {
    'manifest.json': {
      ...manifestV6,
      components: {
        companion: { filename: 'companion.zip' },
      },
    },
    'companion.zip': '',
  });

  it('hydrates a previously-serialized AppPackage instance', async () => {
    const sourcePackage = new AppPackage(testPackage);

    return expect(AppPackage.fromArtifact(await sourcePackage.generateArtifact()))
      .resolves.toMatchObject(sourcePackage);
  });

  describe('when hydrating a previously-serialized source mapped AppPackage', () => {
    async function expectSourceMappedPackageToDeserialize(sourceMaps: SourceMaps) {
      const sourcePackage = new AppPackage({
        ...testPackage,
        sourceMaps,
      });

      return expect(AppPackage.fromArtifact(await sourcePackage.generateArtifact()))
        .resolves.toMatchObject(sourcePackage);
    }

    it('succeeds', () =>
      expectSourceMappedPackageToDeserialize({
        device: {
          higgs: {
            'app/index.js': rawSourceMap,
            'app/test.js': rawSourceMap,
          },
          meson: {
            'app/index.js': rawSourceMap,
          },
        },
        companion: {
          'companion.js': rawSourceMap,
        },
        settings: {
          'settings.js': rawSourceMap,
        },
      }),
    );

    it('succeeds with a component without source map files', () =>
      expectSourceMappedPackageToDeserialize({
        device: {
          higgs: {
            'app/index.js': rawSourceMap,
          },
        },
        companion: {},
      }),
    );

    it('succeeds with a device source map object that has no components', () =>
      expectSourceMappedPackageToDeserialize({ device : {} }));
  });
});

describe('generateArtifact', () => {
  describe('when bundling source map files', () => {
    it('bundles multiple source map files for one component', async () => {
      const sourcePackage = new AppPackage({
        ...testPackage,
        sourceMaps: {
          device: {
            higgs: {
              'app/index.js': rawSourceMap,
              'app/test.js': rawSourceMap,
            },
          },
        },
      });

      const fbaZip = await jszip.loadAsync(await sourcePackage.generateArtifact());
      const manifestJSON = JSON.parse(await fbaZip.file('manifest.json').async('text'));
      const higgsFiles = manifestJSON.sourceMaps.device.higgs;

      await expect(fbaZip.file(higgsFiles['app/index.js']).async('text').then(JSON.parse))
        .resolves.toMatchObject({});

      return expect(fbaZip.file(higgsFiles['app/test.js']).async('text').then(JSON.parse))
        .resolves.toMatchObject({ mappings: 'someMappings' });
    });

    it('bundles companion source map files', async () => {
      const sourcePackage = new AppPackage({
        ...testPackage,
        sourceMaps: {
          device: {
            higgs: {
              'app/index.js': rawSourceMap,
            },
          },
          companion: {
            'companion.js': rawSourceMap,
          },
        },
      });

      const fbaZip = await jszip.loadAsync(await sourcePackage.generateArtifact());
      const manifestJSON = JSON.parse(await fbaZip.file('manifest.json').async('text'));
      const filePath = manifestJSON.sourceMaps.companion['companion.js'];

      return expect(fbaZip.file(filePath).async('text').then(JSON.parse))
        .resolves.toMatchObject({});
    });

    it('bundles settings source map files', async () => {
      const sourcePackage = new AppPackage({
        ...testPackage,
        sourceMaps: {
          device: {
            higgs: {
              'app/index.js': rawSourceMap,
            },
          },
          settings: {
            'settings.js': rawSourceMap,
          },
        },
      });

      const fbaZip = await jszip.loadAsync(await sourcePackage.generateArtifact());
      const manifestJSON = JSON.parse(await fbaZip.file('manifest.json').async('text'));
      const filePath = manifestJSON.sourceMaps.settings['settings.js'];

      return expect(fbaZip.file(filePath).async('text').then(JSON.parse))
        .resolves.toMatchObject({});
    });
  });
});
