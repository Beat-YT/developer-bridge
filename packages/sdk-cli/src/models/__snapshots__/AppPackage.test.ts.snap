// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromArtifact accepts a v5 manifest with a companion 1`] = `
AppPackage {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": Object {
      "data": Array [
        99,
        111,
        109,
        112,
        97,
        110,
        105,
        111,
        110,
      ],
      "type": "Buffer",
    },
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
          ],
          "type": "Buffer",
        },
        "platform": undefined,
      },
    },
  },
  "manifest": Object {
    "appId": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
    "buildId": "0x0123456789abcdef",
    "components": Object {
      "companion": "companion.zip",
      "watch": "device.zip",
    },
    "manifestVersion": 5,
    "platform": Array [
      "HIGGS",
    ],
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "companionApi": "1.0.0",
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromArtifact accepts a v5 manifest with a versioned platform 1`] = `
AppPackage {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "32.1.16+",
        ],
      },
    },
  },
  "manifest": Object {
    "appId": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
    "buildId": "0x0123456789abcdef",
    "components": Object {
      "watch": "device.zip",
    },
    "manifestVersion": 5,
    "platform": Array [
      "HIGGS:32.1.16+",
    ],
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromArtifact accepts a v5 manifest with an unversioned platform 1`] = `
AppPackage {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": undefined,
      },
    },
  },
  "manifest": Object {
    "appId": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
    "buildId": "0x0123456789abcdef",
    "components": Object {
      "watch": "device.zip",
    },
    "manifestVersion": 5,
    "platform": Array [
      "HIGGS",
    ],
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromArtifact accepts a v5 manifest with multiple platforms 1`] = `
AppPackage {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "32.1.16+",
        ],
      },
      "meson": Object {
        "artifact": Object {
          "data": Array [
            100,
            101,
            118,
            105,
            99,
            101,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "1.2.3+",
        ],
      },
    },
  },
  "manifest": Object {
    "appId": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
    "buildId": "0x0123456789abcdef",
    "components": Object {
      "watch": "device.zip",
    },
    "manifestVersion": 5,
    "platform": Array [
      "HIGGS:32.1.16+",
      "MESON:1.2.3+",
    ],
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromArtifact accepts a v6 manifest with a companion 1`] = `
AppPackage {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": Object {
      "data": Array [
        99,
        111,
        109,
        112,
        97,
        110,
        105,
        111,
        110,
      ],
      "type": "Buffer",
    },
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            104,
            105,
            103,
            103,
            115,
          ],
          "type": "Buffer",
        },
        "platform": undefined,
      },
    },
  },
  "manifest": Object {
    "appId": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
    "buildId": "0x0123456789abcdef",
    "components": Object {
      "companion": Object {
        "filename": "companion.zip",
      },
      "watch": Object {
        "higgs": Object {
          "filename": "higgs.zip",
        },
      },
    },
    "manifestVersion": 6,
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "companionApi": "1.0.0",
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromArtifact accepts a v6 manifest with multiple device components 1`] = `
AppPackage {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            104,
            105,
            103,
            103,
            115,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "23.1.5",
          "32.4.16+",
        ],
      },
      "meson": Object {
        "artifact": Object {
          "data": Array [
            109,
            101,
            115,
            111,
            110,
          ],
          "type": "Buffer",
        },
        "platform": Array [
          "32.4.17+",
        ],
      },
    },
  },
  "manifest": Object {
    "appId": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
    "buildId": "0x0123456789abcdef",
    "components": Object {
      "watch": Object {
        "higgs": Object {
          "filename": "higgs.zip",
          "platform": Array [
            "23.1.5",
            "32.4.16+",
          ],
        },
        "meson": Object {
          "filename": "meson.zip",
          "platform": Array [
            "32.4.17+",
          ],
        },
      },
    },
    "manifestVersion": 6,
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromArtifact accepts a v6 manifest with no device component 1`] = `
AppPackage {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "device": Object {},
  },
  "manifest": Object {
    "appId": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
    "buildId": "0x0123456789abcdef",
    "components": Object {
      "companion": Object {
        "filename": "companion.zip",
      },
    },
    "manifestVersion": 6,
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "companionApi": "1.0.0",
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromArtifact accepts a v6 manifest with no platform spec for the device 1`] = `
AppPackage {
  "buildId": "0x0123456789abcdef",
  "components": Object {
    "companion": undefined,
    "device": Object {
      "higgs": Object {
        "artifact": Object {
          "data": Array [
            104,
            105,
            103,
            103,
            115,
            46,
            122,
            105,
            112,
            32,
            99,
            111,
            110,
            116,
            101,
            110,
            116,
            115,
          ],
          "type": "Buffer",
        },
        "platform": undefined,
      },
    },
  },
  "manifest": Object {
    "appId": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
    "buildId": "0x0123456789abcdef",
    "components": Object {
      "watch": Object {
        "higgs": Object {
          "filename": "higgs.zip",
        },
      },
    },
    "manifestVersion": 6,
  },
  "requestedPermissions": undefined,
  "sdkVersion": Object {
    "deviceApi": "1.0.0",
  },
  "sourceMaps": undefined,
  "uuid": "eff9d309-eadd-41d7-9af3-696eafc3fa31",
}
`;

exports[`fromArtifact rejects a manifest that is not JSON 1`] = `[SyntaxError: Unexpected token N in JSON at position 0]`;

exports[`fromArtifact rejects a manifest with unsupported version 1`] = `[Error: Unsupported manifest version 100]`;

exports[`fromArtifact rejects a non-zip file 1`] = `[Error: End of data reached (data length = 0, asked index = 4). Corrupted zip ?]`;

exports[`fromArtifact rejects a v5 manifest which references a nonexistent companion component file 1`] = `[Error: companion.zip not present in zip file]`;

exports[`fromArtifact rejects a v5 manifest which references a nonexistent device component file 1`] = `[Error: doesnotexist.zip not present in zip file]`;

exports[`fromArtifact rejects a v5 manifest with mistyped components 1`] = `[Error: No components listed in manifest.json]`;

exports[`fromArtifact rejects a v5 manifest with no device component 1`] = `[Error: No components listed in manifest.json]`;

exports[`fromArtifact rejects a v5 manifest with no platform descriptor 1`] = `[Error: Missing platform descriptors]`;

exports[`fromArtifact rejects a v6 manifest which references a nonexistent companion component file 1`] = `[Error: companion.zip not present in zip file]`;

exports[`fromArtifact rejects a v6 manifest which references a nonexistent device component file 1`] = `[Error: doesnotexist.zip not present in zip file]`;

exports[`fromArtifact rejects a v6 manifest with mistyped components 1`] = `[Error: No components listed in manifest.json]`;

exports[`fromArtifact rejects a zip file without a manifest 1`] = `[Error: manifest.json not present in zip file]`;
